### ISSUE TITLE
Failing tests with -test.count=10

### ISSUE THREAD

**Issue by pwaller** [acln0/perf#12]
Tested on master just now @ 46e3c14.

```
$ go test -c
$ sudo ./perf.test -test.count=10
--- FAIL: TestCount (0.15s)
    --- FAIL: TestCount/Software (0.00s)
        --- FAIL: TestCount/Software/PageFaults (0.00s)
            count_test.go:104: PERF_EVENT_IOC_DISABLE: bad file descriptor
--- FAIL: TestCount (0.12s)
    --- FAIL: TestCount/Software (0.00s)
        --- FAIL: TestCount/Software/PageFaults (0.00s)
            count_test.go:104: PERF_EVENT_IOC_DISABLE: bad file descriptor
--- FAIL: TestReadRecord (0.24s)
    --- FAIL: TestReadRecord/Comm (0.05s)
        record_test.go:635: got context deadline exceeded, want valid record
--- FAIL: TestCount (0.11s)
    --- FAIL: TestCount/Software (0.00s)
        --- FAIL: TestCount/Software/PageFaults (0.00s)
            count_test.go:104: PERF_EVENT_IOC_DISABLE: bad file descriptor
    --- FAIL: TestCount/IoctlAndCountIDsMatch (0.02s)
        count_test.go:368: didn't see a page fault
--- FAIL: TestCount (0.09s)
    --- FAIL: TestCount/Software (0.00s)
        --- FAIL: TestCount/Software/PageFaults (0.00s)
            count_test.go:104: PERF_EVENT_IOC_DISABLE: bad file descriptor
    --- FAIL: TestCount/IoctlAndCountIDsMatch (0.02s)
        count_test.go:368: didn't see a page fault
--- FAIL: TestReadRecord (0.24s)
    --- FAIL: TestReadRecord/Comm (0.05s)
        record_test.go:635: got context deadline exceeded, want valid record
--- FAIL: TestCount (0.10s)
    --- FAIL: TestCount/Software (0.01s)
        --- FAIL: TestCount/Software/PageFaults (0.01s)
            count_test.go:104: PERF_EVENT_IOC_DISABLE: bad file descriptor
    --- FAIL: TestCount/IoctlAndCountIDsMatch (0.02s)
        count_test.go:368: didn't see a page fault
--- FAIL: TestReadRecord (0.25s)
    --- FAIL: TestReadRecord/Comm (0.05s)
        record_test.go:635: got context deadline exceeded, want valid record
--- FAIL: TestCount (0.09s)
    --- FAIL: TestCount/Software (0.00s)
        --- FAIL: TestCount/Software/PageFaults (0.00s)
            count_test.go:104: PERF_EVENT_IOC_DISABLE: bad file descriptor
    --- FAIL: TestCount/IoctlAndCountIDsMatch (0.01s)
        count_test.go:368: didn't see a page fault
--- FAIL: TestReadRecord (0.27s)
    --- FAIL: TestReadRecord/Comm (0.05s)
        record_test.go:635: got context deadline exceeded, want valid record
--- FAIL: TestCount (0.09s)
    --- FAIL: TestCount/Software (0.01s)
        --- FAIL: TestCount/Software/PageFaults (0.01s)
            count_test.go:104: PERF_EVENT_IOC_DISABLE: bad file descriptor
    --- FAIL: TestCount/IoctlAndCountIDsMatch (0.02s)
        count_test.go:368: didn't see a page fault
--- FAIL: TestReadRecord (0.30s)
    --- FAIL: TestReadRecord/Comm (0.05s)
        record_test.go:635: got context deadline exceeded, want valid record
--- FAIL: TestCount (0.09s)
    --- FAIL: TestCount/Software (0.01s)
        --- FAIL: TestCount/Software/PageFaults (0.01s)
            count_test.go:104: PERF_EVENT_IOC_DISABLE: bad file descriptor
    --- FAIL: TestCount/IoctlAndCountIDsMatch (0.02s)
        count_test.go:368: didn't see a page fault
--- FAIL: TestReadRecord (0.26s)
    --- FAIL: TestReadRecord/Comm (0.06s)
        record_test.go:635: got context deadline exceeded, want valid record
--- FAIL: TestCount (0.10s)
    --- FAIL: TestCount/Software (0.00s)
        --- FAIL: TestCount/Software/PageFaults (0.00s)
            count_test.go:104: PERF_EVENT_IOC_DISABLE: bad file descriptor
    --- FAIL: TestCount/IoctlAndCountIDsMatch (0.02s)
        count_test.go:368: didn't see a page fault
--- FAIL: TestReadRecord (0.28s)
    --- FAIL: TestReadRecord/Comm (0.05s)
        record_test.go:635: got context deadline exceeded, want valid record
FAIL
```

**Comment by acln0** [comment]
They fail with any count > 1, apparently. Odd.

Investigating.

**Comment by acln0** [comment]
I disabled `TestReadRecord/Comm` in the `x/sys/unix` variant because it was flaky. I will note that it re-execs the test binary, so perhaps there is some kind of bad interaction there. When I skip the `Comm` test, `TestCount/Software` fails at higher counts too, but because it doesn't see any page faults anymore at a certain run, but at least it's not a mysterious `EBADFD` anymore.

Suggestions for how to more reliably produce a page fault are welcome. In the meantime, I'll disable `TestReadRecord/Comm` here too.

**Comment by acln0** [comment]
As of https://github.com/acln0/perf/commit/95ba6cff1532abf37e077479917d5f36d48a817e, (where I managed to mess up git history, embarrassingly so), `TestCount/Software` should be reliable.

Anecdotally, it seems like disabling all tests which re-exec the test binary causes running with `-count=n` to work fine. Once I confirm that this is indeed the truth, I plan to re-purpose this bug to investigate why that is the case.

Perhaps some file descriptor is not `CLOEXEC` when it should be, and the child closes or writes to a bad file descriptor somehow. No idea.

**Comment by pwaller** [comment]
I think this is fixed by #25; the bad file descriptors are again caused by the fact that both the perf code and os.NewFile were calling close on the file descriptors.
